### ISSUE TITLE
OpenConfig implementation on IOS-XRv BGP neighbor restart-timer is different

### ISSUE THREAD

**Issue by jwbensley** [CiscoDevNet/openconfig-getting-started#5]
NB: This is the gRPC client "grpc_cfg.py" I am using based off of the public Cisco example: https://gist.github.com/jwbensley/cd2f5bfacee321328ca638b5b5523bd9

NB: This is using the IOS-XRv 6.2.1 stable vagrant image from https://devhub.cisco.com/artifactory/appdevci-release/XRv64/6.2.1/ dated 17th Oct 2016

OpenConfig defines the BGP neighbor restart timer (when the maximum number of prefixes allowed has been received) in seconds as a 64 bit decimal figure. IOS-XRv expects an integer that is used to specify minutes.

openconfig-bgp-common-multiprotocol.yang defines the following:

```
  // Config groupings for common groups
  grouping bgp-common-mp-all-afi-safi-common-prefix-limit-config {
    description
      "Configuration parameters relating to prefix-limits for an
      AFI-SAFI";

    leaf max-prefixes {
      type uint32;
      description
        "Maximum number of prefixes that will be accepted
        from the neighbour";
    }

    leaf shutdown-threshold-pct {
      type oc-types:percentage;
      description
        "Threshold on number of prefixes that can be received
        from a neighbour before generation of warning messages
        or log entries. Expressed as a percentage of
        max-prefixes";
    }

    leaf restart-timer {
      type decimal64 {
        fraction-digits 2;
      }
      units "seconds";
      description
        "Time interval in seconds after which the BGP session
        is re-established after being torn down due to exceeding
        the max-prefix limit.";
    }
  }
```

This is the example JSON config:

```
bensley@ubuntu-laptop:~/Python/pyb_oc/ios-xr-grpc-python-master/examples$ cat snips/broken_restart_timer.json 
{
    "openconfig-bgp:bgp": {
        "neighbors": {
            "neighbor": [
                {
                    "neighbor-address": "193.37.144.1", 
                    "config": {
                        "neighbor-address": "193.37.144.1", 
                        "description": "i3b gmbh", 
                        "peer-as": 39912, 
                        "peer-group": "public_v4_peers"
                    }, 
                    "afi-safis": {
                        "afi-safi": [
                            {
                                "afi-safi-name": "openconfig-bgp-types:ipv4-unicast", 
                                "config": {
                                    "enabled": true, 
                                    "afi-safi-name": "openconfig-bgp-types:ipv4-unicast"
                                }, 
                                "ipv4-unicast": {
                                    "prefix-limit": {
                                        "config": {
                                            "shutdown-threshold-pct": 90, 
                                            "restart-timer": 10.00, 
                                            "max-prefixes": 200
                                        }
                                    }
                                }
                            }
                        ]
                    }
                }
            ]
        }
    }
}
```

Note the restart-timer value is 10.00. This produces the following error from IOS-XRv:

`{u'cisco-grpc:errors': {u'error': [{u'error-type': u'protocol', u'error-severity': u'error', u'error-tag': u'bad-element', u'error-path': u"bgp:ns1:bgp/neighbors/neighbor[neighbor-address = '193.37.144.1']/afi-safis/afi-safi[afi-safi-name = 'ipv4-unicast']/ipv4-unicast/prefix-limit/config/restart-timer", u'error-info': {u'bad-element': u'restart-timer'}}]}}`

Setting it to 10 works fine.

```
Config before merge:

{
    "Cisco-IOS-XR-ipv4-bgp-cfg:bgp": {
        "instance": [
            {
                "instance-as": [
                    {
                        "four-byte-as": [
                            {
                                "as": 65001, 
                                "bgp-running": [
                                    null
                                ], 
                                "default-vrf": {
                                    "global": {
                                        "global-afs": {
                                            "global-af": [
                                                {
                                                    "af-name": "ipv4-unicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }, 
                                                {
                                                    "af-name": "ipv6-unicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }
                                            ]
                                        }, 
                                        "router-id": "1.1.0.10"
                                    }, 
                                    "bgp-entity": {
                                        "neighbor-groups": {
                                            "neighbor-group": [
                                                {
                                                    "neighbor-group-afs": {
                                                        "neighbor-group-af": [
                                                            {
                                                                "af-name": "ipv4-unicast", 
                                                                "activate": [
                                                                    null
                                                                ], 
                                                                "route-policy-in": "public_v4_peering_ingress", 
                                                                "route-policy-out": "public_v4_peering_egress"
                                                            }
                                                        ]
                                                    }, 
                                                    "create": [
                                                        null
                                                    ], 
                                                    "neighbor-group-name": "public_v4_peers"
                                                }, 
                                                {
                                                    "neighbor-group-afs": {
                                                        "neighbor-group-af": [
                                                            {
                                                                "af-name": "ipv6-unicast", 
                                                                "activate": [
                                                                    null
                                                                ], 
                                                                "route-policy-in": "public_v6_peering_ingress", 
                                                                "route-policy-out": "public_v6_peering_egress"
                                                            }
                                                        ]
                                                    }, 
                                                    "create": [
                                                        null
                                                    ], 
                                                    "neighbor-group-name": "public_v6_peers"
                                                }
                                            ]
                                        }
                                    }
                                }
                            }
                        ], 
                        "as": 0
                    }
                ], 
                "instance-name": "default"
            }
        ]
    }
}
Config after merge:

{
    "Cisco-IOS-XR-ipv4-bgp-cfg:bgp": {
        "instance": [
            {
                "instance-as": [
                    {
                        "four-byte-as": [
                            {
                                "as": 65001, 
                                "bgp-running": [
                                    null
                                ], 
                                "default-vrf": {
                                    "global": {
                                        "global-afs": {
                                            "global-af": [
                                                {
                                                    "af-name": "ipv4-unicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }, 
                                                {
                                                    "af-name": "ipv6-unicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }
                                            ]
                                        }, 
                                        "router-id": "1.1.0.10"
                                    }, 
                                    "bgp-entity": {
                                        "neighbor-groups": {
                                            "neighbor-group": [
                                                {
                                                    "neighbor-group-afs": {
                                                        "neighbor-group-af": [
                                                            {
                                                                "af-name": "ipv4-unicast", 
                                                                "activate": [
                                                                    null
                                                                ], 
                                                                "route-policy-in": "public_v4_peering_ingress", 
                                                                "route-policy-out": "public_v4_peering_egress"
                                                            }
                                                        ]
                                                    }, 
                                                    "create": [
                                                        null
                                                    ], 
                                                    "neighbor-group-name": "public_v4_peers"
                                                }, 
                                                {
                                                    "neighbor-group-afs": {
                                                        "neighbor-group-af": [
                                                            {
                                                                "af-name": "ipv6-unicast", 
                                                                "activate": [
                                                                    null
                                                                ], 
                                                                "route-policy-in": "public_v6_peering_ingress", 
                                                                "route-policy-out": "public_v6_peering_egress"
                                                            }
                                                        ]
                                                    }, 
                                                    "create": [
                                                        null
                                                    ], 
                                                    "neighbor-group-name": "public_v6_peers"
                                                }
                                            ]
                                        }, 
                                        "neighbors": {
                                            "neighbor": [
                                                {
                                                    "neighbor-address": "193.37.144.1", 
                                                    "remote-as": {
                                                        "as-xx": 0, 
                                                        "as-yy": 39912
                                                    }, 
                                                    "neighbor-group-add-member": "public_v4_peers", 
                                                    "description": "i3b gmbh", 
                                                    "neighbor-afs": {
                                                        "neighbor-af": [
                                                            {
                                                                "af-name": "ipv4-unicast", 
                                                                "activate": [
                                                                    null
                                                                ], 
                                                                "maximum-prefixes": {
                                                                    "discard-extra-paths": false, 
                                                                    "warning-percentage": 90, 
                                                                    "warning-only": false, 
                                                                    "prefix-limit": 200, 
                                                                    "restart-time": 10
                                                                }
                                                            }
                                                        ]
                                                    }
                                                }
                                            ]
                                        }
                                    }
                                }
                            }
                        ], 
                        "as": 0
                    }
                ], 
                "instance-name": "default"
            }
        ]
    }
}
```

We can see the config has been applied below. However in IOS-XRv we have configured a 10 minute restart time when a 10.0 second restart timer was desired:

```
RP/0/RP0/CPU0:ios#show run router bgp
Wed Nov  9 17:18:42.167 UTC
router bgp 65001
 bgp router-id 1.1.0.10
 address-family ipv4 unicast
 !
 address-family ipv6 unicast
 !
 neighbor-group public_v4_peers
  address-family ipv4 unicast
   route-policy public_v4_peering_ingress in
   route-policy public_v4_peering_egress out
  !
 !
 neighbor-group public_v6_peers
  address-family ipv6 unicast
   route-policy public_v6_peering_ingress in
   route-policy public_v6_peering_egress out
  !
 !
 neighbor 193.37.144.1
  remote-as 39912
  use neighbor-group public_v4_peers
  description i3b gmbh
  address-family ipv4 unicast
   maximum-prefix 200 90 restart 10
  !
 !
!

RP/0/RP0/CPU0:ios#conf t
Wed Nov  9 17:18:52.303 UTC
RP/0/RP0/CPU0:ios(config)#router bgp 65001 neighbor 193.37.144.1 address-family ipv4 unicast
RP/0/RP0/CPU0:ios(config-bgp-nbr-af)#maximum-prefix 200 90 restart ?
  <1-65535>  Time interval (min) after which peering session will be reestablished

RP/0/RP0/CPU0:ios#show bgp ipv4 unicast neighbors 193.37.144.1 | inc 10
Wed Nov  9 17:20:45.309 UTC
  Threshold for warning message 90%, restart interval 10 min
```

Is the OpenConfig model for this not correctly implemented in IOS-XRv 6.2.1?

**Comment by 111pontes** [comment]
This XR issue is being tracked with CSCvc09402.

**Comment by jwbensley** [comment]
@111pontes  Thanks for opening the bug case! I'll wait to here from you when its fixed :)

**Comment by 111pontes** [comment]
Fixed. Expected to be released in XR 6.2.1.
